src/jsonrpc/miner_jsonrpc_ledger.erl

```diff
@@ -189,3 +211,40 @@ format_ledger_validator(Addr, Val, Ledger, Height) ->
         <<"performance_penalty">> => Perf,
         <<"total_penalty">> => TotalPenalty
     }.
+
+-spec format_region_parameters([blockchain_region_param_v1:region_param_v1()]) ->
+    [map()].
+format_region_parameters(RegionParameters) ->
+    [ format_region_parameter(RegionParameter) || RegionParameter <- RegionParameters ].
+
+-spec format_region_parameter(blockchain_region_param_v1:region_param_v1()) ->
+    map().
+format_region_parameter(RegionParam) ->
+    ChannelFrequencyHz = blockchain_region_param_v1:channel_frequency(RegionParam),
+    BandwidthHz = blockchain_region_param_v1:bandwidth(RegionParam),
+    MaxEIRPcBm = blockchain_region_param_v1:max_eirp(RegionParam),
+    Spreading = blockchain_region_param_v1:spreading(RegionParam),
+    SpreadingObject = format_spreading(Spreading),
+    #{
+        <<"channel_frequency_hz">> => ChannelFrequencyHz,
+        <<"bandwidth_hz">> => BandwidthHz,
+        <<"max_eirp_cbm">> => MaxEIRPcBm,
+        <<"spreading">> => SpreadingObject
+    }.
```

Comment:
for typespecs on functions that return relatively small maps that don't change frequently (or complex tuples for that matter) do we have style guidelines for when we might want a function like this to define a spec like
```erlang
-spec format_region_parameter(blockchain_region_param_v1:region_param_v1()) -> #{binary() => any()}
```
or would we generally prefer to limit this type of specificity to functions that return records?
In this case, I can see the possibility of the `undefined` atom as potential return values for the fields making the map cumbersome so if there's no sensible defaults for region params it probably doesn't make sense for this map.

Reply:
I'm still new to Erlang type specs and would love more guidance here. I didn't know how deep to go, nor how to specify a map that had specific keys and the types for those keys.

Reply:
It's largely a matter of preference and documentation if we're not consistently running dialyzer, particularly as a regular CI check but the example above is a middle ground between `map()` and a fully explicit type definition. With a map that has static contents you could then guard even further against changes that deviate from that spec with `#{<<"channel_frequency_hz">> => undefined | non_neg_integer(), <<"bandwidth_hz">> => undefined |non_neg_integer(), <<"max_eirp_cbm">> => undefined | non_neg_integer(), <<"spreading">> => undefined | blockchain_region_spreading_v1:region_spreading_v1()}` but the possibility of those `undefined`s makes that level of specificity less valuable in my opinion

Reply:
I like @jeffgrunewald 's middle ground.. Over-speccing, generally takes a lot of time for little benefit

Reply:
I'm not sure what the middle-ground would look like. How is this?

```erlang
-type lora_spreading_spec() :: #{
        <<"factor">> => binary(),
        <<"max_packet_size">> => non_neg_integer()
    }.
-type lora_channel_spec() :: #{
        <<"channel_frequency_hz">> => non_neg_integer(),
        <<"bandwidth_hz">> => non_neg_integer(),
        <<"max_eirp_cbm">> => integer(),
        <<"spreading">> => [lora_spreading_spec()]
    }.
```

Reply:
Oh, I see the middle ground now. Well, as you can see, a fully-spec'd type for these isn't that bad.

Reply:
+1 on middle ground.

Reply:
Well, my example above apparently isn't a legal type spec. I don't know how to spec a map with specific keys; I can't figure it out from the general BNF for type specs.

Reply:
That’s right; I recall you can only use explicit map keys in a spec if they’re atoms. `binary()` is probably the most specific binary-type keys would support

Reply:
Ok, how about the latest change? I introduced a type called `json_object()` which is nothing more than `#{ binary() => any() }.`